include/sdf/ParserConfig.hh

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef SDF_PARSER_CONFIG_HH_
+#define SDF_PARSER_CONFIG_HH_
+
+#include <functional>
+#include <list>
+#include <map>
+#include <string>
+
+#include <sdf/Error.hh>
+#include "sdf/sdf_config.h"
+#include "sdf/system_util.hh"
+
+namespace sdf
+{
+inline namespace SDF_VERSION_NAMESPACE
+{
+// Forward declare private data class.
+class ParserConfigPrivate;
+
+/// This class contains configuration options for the libsdformat parser.
+///
+/// The configuration options include:
+///   * A callback function used to locate a file from a given URI
+///   * A map from URI schemes to search directories
+///
+/// For backward compatibility, the functions sdf::setFindCallback() and
+/// sdf::addURIPath() update a singleton ParserConfig object, which can be
+/// retrieved by ParserConfig::DefaultConfig().
+///
+/// The functions sdf::readFile(), sdf::readString(), and \ref Root::Load
+/// "sdf::Root::Load()" have overloads that take a ParserConfig object. If
+/// the ParserConfig object is omitted, these functions will use the singleton
+/// ParserConfig object.
+///
+/// Example:
+/// To set an additional URI scheme search directory without affecting the
+/// default config,
+///
+/// \code{.cpp}
+///   // Copy the default config
+///   sdf::ParserConfig config = ParserConfig::Default();
+///
+///   // Add the new scheme to the config
+///   config.AddURIPath("newScheme://", "path/to/directory");
+///
+///   // Use the new config when loading a new SDFormat file
+///   sdf::Root root;
+///   root.Load("path/to/file.sdf", config);
+/// \endcode
+class SDFORMAT_VISIBLE ParserConfig
+{
+  /// type alias for the map from URI scheme to search directories
+  public: using SchemeToPathMap =
+          std::map<std::string, std::list<std::string> >;
```

Comment:
Out of curiosity, why a std::list instead of std::vector?

Reply:
This was copied form https://github.com/osrf/sdformat/blob/master/src/SDF.cc#L40 but I also think `std::vector` makes more sense. Changed in https://github.com/osrf/sdformat/pull/439/commits/f9e3ef294641cf7f963b688789b45684582c4b02